Documentation - Centralized Telescope: Main (#101)
cardano-scaling/alba: docs/rustdoc/centralized_telescope/main.md

# Telescope - Construction with Bounded DFS

In scenarios where $n_p$ is small, the parameters given for prehashed construction (relative to the security parameter, $\lambda$) are not optimal, resulting in a reduced probability of constructing a valid proof in a single attempt.
For large $n_p$, the rapid growth in potential proof tuples ensures valid ones can be found efficiently, allowing the prehashed construction to work seamlessly.
In contrast, small $n_p$ limits the search space, making the previous parameters inadequate.
To address this, the **construction with bounded DFS** expands on the prehashed version with *retries*, *randomized indexing*, and *bounding* the DFS.

## Overview
- When $n_p$ is large, the rapid growth in potential proof tuples increases the chances of finding a valid proof, making construction easier.

> - For small $n_p$, limited elements reduce the probability of finding a valid proof in a single attempt.

Reviewer: The real reason is that when n_p is small, the arrangement of balls into bins is far from uniform. On average each bin still gets exactly one ball, but it's likely that many bins will be empty and some will have many elements. Turns out it's better to have many bins with one ball than one bin with many balls.

Reviewer: When n_p is small, we still can use the scheme with prehashing, but the completeness error is moderate instead of tiny. By introducing retries, the completeness error is decreased.

- The generalized scheme compensates by using multiple retries indexed by $v$, where each retry applies fresh randomization via hash functions.
- A bounded DFS search with a predefined step limit ensures computational efficiency and deterministic worst-case running time.

### Core components
This generalized construction uses the same key parameters as prehashed construction but introduces:
- Retries with index $v$: 
  - The prover retries the proof generation process up to $r$ times.
  - Each retry uses a different retry counter $v$, where $v \in \[1, r\]$, to diversify the search process.

> - Hash-based binning:

Reviewer 2: ```suggestion
- Seeded binning:
```

>   - Elements in $S_p$ are prehashed into bins using $H_0(v, \cdot) where v \in \[1,r\]$, grouping elements based on their hash values. We prehashed using the retry counter to get different bins for each repetition, reducing the risk of badly distributed bins.

Reviewer 2: ```suggestion
  - Elements in $S_p$ are prehashed into bins using $H_0(v, \cdot)\ where\ v \in \[1,r\]$, grouping elements based on their hash values. We prehashed using the retry counter to get different bins for each repetition, reducing the risk of badly distributed bins.
```

>   - Elements in $S_p$ are prehashed into bins using $H_0(v, \cdot) where v \in \[1,r\]$, grouping elements based on their hash values. We prehashed using the retry counter to get different bins for each repetition, reducing the risk of badly distributed bins.

Reviewer: ```suggestion
  - Elements in $S_p$ are prehashed into bins using $H_0(v, \cdot)$ where $v \in \[1,r\]$, grouping elements based on their hash values. We prehashed using the retry counter to get different bins for each repetition, reducing the risk of badly distributed bins.
```

  - A depth-first search explores valid proof sequences of size $u$, using the bins for efficient lookup.
  - The total number of steps is bounded by a limit to ensure efficiency.

## Protocol
1. **Initialization**:
   - The prover begins by preparing for up to $r$ retries. 
   - Each retry introduces a unique retry counter $v \in \[1, r\]$ to diversify the search space.
   - Each retry is an independent attempt to construct a valid proof.
2. **Prehashing**:
   - For each retry, elements in $S_p$ are hashed using the hash function $H_0(v, s)$.
   - This process groups elements based on their hash values, reducing the search space by allowing DFS to focus only on relevant elements.
3. **Exploring starting points**:
   - The prover iterates over all possible starting indices $t \in \[1, d\]$.
   - For each $t$, the prover begins constructing a proof sequence $(t, s_1, ..., s_u)$, starting with $t$ and extending the sequence using elements from $S_p$ within their corresponding bins.
4. **Bounded DFS**:
   - A bounded DFS search is used to construct the sequence $(t, s_1, ..., s_u)$, with a shared step limit $B$ applied across all starting points $t$.
   - At each step of DFS:

>      - The algorithm search a new element for the current sequence $(t, s_1, ..., s_k)$ in bin numbered $H_1(v, t, s_1, ..., s_k)$.

Reviewer: ```suggestion
     - The algorithm searches a new element for the current sequence $(t, s_1, ..., s_k)$ in bin numbered $H_1(v, t, s_1, ..., s_k)$.
```

     - If the step limit $B$ is reached or no valid extension exists, the DFS backtracks and explores another retry $r$, restarting the process with a new partitioning of elements. 
     - Otherwise, the DFS updates the step limit and then recursively calls itself with the extended tuple.
5. **Validation**:
   - When the DFS has constructed a sequence of $u$ elements $(t, s_1, ..., s_u)$, it verifies whether the full sequence satisfies the final condition using $H_2(v, t, s_1, ..., s_u)$.
   - If the sequence passes this check, it is accepted as a valid proof, and the prover outputs it.
6. **Retry Mechanism**:
   - If no valid proof is found for a given $t$ or if the step limit $B$ is exhausted for all $t$, the prover moves to the next retry by incrementing $v$.
   - With the new $v$, the hash function $H_0(v, s)$ is applied to organize elements of $S_p$ into bins, and the process resumes from step 3.
7. **Completion**:
   - If a valid proof is found in any retry, the prover outputs the proof immediately.

>    - If all $r$ retries are exhausted without finding a valid proof, the process terminates, and the prover outputs $\bot$.

Reviewer: Will there be pseudocode like in previous docs?

Author: Yes. I will add them in separate PRs.

# Telescope - Construction with Bounded DFS

> 

Reviewer: Is this file in `docs/rustdoc/centralized_telescope/` on purpose as opposed to `docs/rustdoc/` like for the basic and prehashed docs?

Author: Centralized telescope will include several files. This is the main page, parameter setup and other algorithm explanations will have their own pages as opposed to basic and prehashed docs. So, I decided to store them in a folder.

Reviewer: Ok, I see.

Reviewer: Since this doc is in a different location, maybe add a link to the prehashed version doc when you mention it at the beginning?

In scenarios where $n_p$ is small, the parameters given for prehashed construction (relative to the security parameter, $\lambda$) are not optimal, resulting in a reduced probability of constructing a valid proof in a single attempt.
For large $n_p$, the rapid growth in potential proof tuples ensures valid ones can be found efficiently, allowing the prehashed construction to work seamlessly.
In contrast, small $n_p$ limits the search space, making the previous parameters inadequate.

> To address this, the **construction with bounded DFS** expands on the prehashed version with *retries*, *randomized indexing*, and *bounding* the DFS.

Reviewer: what is randomized indexing?

Author: Prehashing with retry counter $r$.

Reviewer 2: Something like `prehashing randomization` mightbe clearer then?

Reviewer: I would omit it altogether. "Retries" already covers it, the counter $r$ is an implementation detail.

## Overview
- When $n_p$ is large, the rapid increase in potential proof tuples enhances the chances of finding a valid proof, simplifying the construction process.

> - For small $n_p$, the element distribution across bins becomes skewed, some bins may contain multiple elements while others remain empty.

Reviewer 2: ```suggestion
- For small $n_p$, the element distribution across bins is more likely to be skewed, some bins may contain multiple elements while others remain empty.
```

  - It significantly lowers the completeness error and improves success rates.
- A bounded DFS search with a predefined step limit ensures computational efficiency while maintaining a deterministic worst-case runtime.

### Core components
This generalized construction uses the same key parameters as prehashed construction but introduces:
- Retries with index $v$: 
  - The prover retries the proof generation process up to $r$ times.
  - Each retry uses a different retry counter $v$, where $v \in \[1, r\]$, to diversify the search process.
- Seeded binning:
  - Elements in $S_p$ are prehashed into bins using $H_0(v, \cdot)$ where $v \in \[1,r\]$, grouping elements based on their hash values. We prehashed using the retry counter to get different bins for each repetition, reducing the risk of badly distributed bins.
  - This process limits the search space for DFS, making it more efficient.
- Bounded DFS:
  - A depth-first search explores valid proof sequences of size $u$, using the bins for efficient lookup.
  - The total number of steps is bounded by a limit to ensure efficiency.

## Protocol
1. **Initialization**:
   - The prover begins by preparing for up to $r$ retries. 

>    - Each retry introduces a unique retry counter $v \in \[1, r\]$ to diversify the search space.

Reviewer: should we use [r] here like in the prehashed doc?

Author: You are right; I will have a second look at both docs to see if there is an occurrence I missed.

   - Each retry is an independent attempt to construct a valid proof.
2. **Prehashing**:
   - For each retry, elements in $S_p$ are hashed using the hash function $H_0(v, s)$.
   - This process groups elements based on their hash values, reducing the search space by allowing DFS to focus only on relevant elements.
3. **Exploring starting points**:

>    - The prover iterates over all possible starting indices $t \in \[d\]$.

Reviewer 2: What was the naming that we agreed on for `t`? I do not think it was " starting indices".

   - For each $t$, the prover begins constructing a proof sequence $(v, t, s_1, \ldots, s_u)$, starting with $t$ and extending the sequence using elements from $S_p$ within their corresponding bins.
4. **Bounded DFS**:

>    - A bounded DFS search is used to construct the sequence $(v, t, s_1, \ldots, s_u)$, with a shared step limit $B$ applied across all starting points $t$.

Reviewer 2: You can add, to highlight, that the bound is reset for each `v`.

   - At each step of DFS:
     - The algorithm searches a new element for the current sequence $(v, t, s_1, \ldots, s_k)$ in bin numbered $H_1(v, t, s_1, \ldots, s_k)$.

>      - If the step limit $B$ is reached or no valid extension exists, the DFS backtracks and explores another retry $r$, restarting the process with a new partitioning of elements. 

Reviewer 2: ```suggestion
     - If the step limit $B$ is reached or no valid extension exists, the DFS backtracks and explores another retry $v \in \[r\]$ , restarting the process with a new partitioning of elements. 
```

     - Otherwise, the DFS updates the step limit and then recursively calls itself with the extended tuple.
5. **Validation**:

>    - When the DFS has constructed a sequence of $u$ elements $(v, t, s_1, \ldots, s_u)$, it verifies whether the full sequence satisfies the final condition using $H_2(v, t, s_1, \ldots, s_u)$.

Reviewer 2: ```suggestion
   - When the DFS has constructed a sequence of $u$ elements $(v, t, s_1, \ldots, s_u)$, it verifies whether the full sequence satisfies the final oracle check, $H_2(v, t, s_1, \ldots, s_u) == 1$.
```
